bind_test.go

```diff
@@ -0,0 +1,229 @@
+		name := fmt.Sprintf("%s.example", alg)
+		args := append(baseArgs, alg, "-b", bits, name)
+		cmd := exec.Command("dnssec-keygen", args...)
+		var stdout, stderr bytes.Buffer
+		cmd.Stdout = &stdout
+		cmd.Stderr = &stderr
+		if err := cmd.Run(); err != nil {
+			log.Printf(`dnssec-keygen failed: %s"
+args: %s
+stderr:
+%s`, err, strings.Join(args, " "), stderr.String())
+		} else {
+			filename := filepath.Join(tempdir, strings.TrimSpace(stdout.String())+".key")
+			contents, err := ioutil.ReadFile(filename)
+			if err != nil {
+				panic(err)
+			}
+			fields := strings.Split(string(contents), " ")
+			secret := strings.TrimSpace(fields[len(fields)-1])
+			keys[alg] = secret
+		}
+	}
+	return keys
+}
+
+func bindBuildZone(tempdir string, dnssecKeys map[uint8]string, hmacKeys map[string]string) (string, error) {
+	zonefilePath := filepath.Join(tempdir, "example.db")
+	signedzonefilePath := zonefilePath + ".signed"
+	zonefile, err := os.Create(zonefilePath)
+	if err != nil {
+		return "", err
+	}
+	_, err = zonefile.WriteString(`
+$TTL 60
+@ SOA mname rname 10 20 30 40 50
+    NS localhost
+localhost A 127.0.0.1
+    `)
+	if err != nil {
+		return "", err
+	}
+	for _, k := range dnssecKeys {
+		_, err = zonefile.WriteString(fmt.Sprintf("\n$INCLUDE %s.key\n", filepath.Join(tempdir, k)))
+		if err != nil {
+			return "", err
+		}
+	}
+	if err = zonefile.Close(); err != nil {
+		return "", err
+	}
+	signArgs := []string{"-z", "-S", "-N", "keep", "-o", "example", "-K", tempdir, "-d", tempdir, "-f", signedzonefilePath, zonefilePath}
+	cmd := exec.Command("dnssec-signzone", signArgs...)
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+	return signedzonefilePath, nil
+}
+
+func bindStart(tempdir string) (*os.Process, chan struct{}, error) {
+	dnssecKeys := bindGenerateDnssecKeys(tempdir)
+	hmacKeys := bindGenerateHmacKeys(tempdir)
+	zonefile, err := bindBuildZone(tempdir, dnssecKeys, hmacKeys)
+	if err != nil {
+		return nil, nil, err
+	}
+	conffilePath := filepath.Join(tempdir, "bind.conf")
+	conffile, err := os.Create(conffilePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	var updatePolicy, hmacKeyStatements string
+	for alg, secret := range hmacKeys {
+		updatePolicy += fmt.Sprintf(`grant "%s.example." selfwild any;`, alg)
+		hmacKeyStatements += fmt.Sprintf(`key %s.example. { algorithm %s; secret "%s"; };
+`, alg, alg, secret)
+	}
+	_, err = conffile.WriteString(fmt.Sprintf(`
+controls { };
+options {
+    allow-query-cache { none; };
+    allow-query { any; };
+    allow-transfer { any; };
+    notify no;
+    recursion no;
+    listen-on { 127.0.0.1; };
+};
+zone "example." {
+    type master;
+    file "%s";
+    allow-transfer { any; };
+    update-policy { %s };
+};
+%s
+`, zonefile, updatePolicy, hmacKeyStatements))
+	if err != nil {
+		return nil, nil, err
+	}
+	cmd := exec.Command("named", "-p", "55033", "-g", "-c", conffilePath)
```

Comment:
fix port is going to give is trouble

Reply:
Yep. I put a var on `bindHarnessInfo` for the port but ran out of time to grab a free port, and forgot to add a todo.